pkg/client/image_test.go

```diff
@@ -0,0 +1,323 @@
+package client_test
+
+// TODO: Add failure states to be tested
+
+import (
+	"context"
+
+	v12 "github.com/acorn-io/acorn/pkg/apis/api.acorn.io/v1"
+	"github.com/acorn-io/acorn/pkg/client"
+	"github.com/acorn-io/acorn/pkg/labels"
+	"github.com/acorn-io/acorn/pkg/mocks"
+	scheme2 "github.com/acorn-io/acorn/pkg/scheme"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	// may want to use envtest instead
+	"testing"
+
+	testcontrollerclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+type defaultClientImagesConfig struct {
+	images      v12.ImageList
+	projectName string
+	namespace   string
+}
+
+func createMockedDefaultClientImageLister(t *testing.T, config defaultClientImagesConfig) (client.DefaultClient, error) {
```

Comment:
Now that I look at this, I'm not sure this is needed. Why doesn't the `gomock` package work here?

Reply:
It does work. Didn't want to mock a k8's library when k8 provided their own fake library for it. Figured I'd run into random versioning issues or something else.

I see we are now mocking the server k8s library, but I feel using k8's fake library is more robust. I could be completely wrong.

Reply:
We've discussed this in a few different places. Let's not introduce k8s's fake library if we can avoid it.

It seems fine and would address the use case, but I want to limit the number of different ways we mock things. @ibuildthecloud pointed out at that baaah has a test client that can accomplish this and it is what is used behind the scenes for all our yaml-based tests.

Here is an example usage:
https://github.com/acorn-io/acorn/blob/f6ce48cd568bc3623e3e343f62e0811a83672945/pkg/ref/expr_test.go#L216-L219

Reply:
Can't currently baaah's client to work as kclient.withWatch because it doesn't implement Watch, so can't use that as the base client underneath defaultclient.

Is it worthwhile for me to modify baaah to be able to use it, or to use what is currently written?